akka-stream-tests/src/test/scala/akka/stream/scaladsl/SinkForeachAsyncSpec.scala

```diff
@@ -0,0 +1,128 @@
+import akka.stream.testkit.StreamSpec
+import akka.stream.testkit.Utils._
+import akka.testkit.{ TestLatch, TestProbe }
+
+import scala.concurrent.duration._
+import scala.concurrent.{ Await, ExecutionContext, Future }
+
+class SinkForeachAsyncSpec extends StreamSpec {
+
+  implicit val materializer = ActorMaterializer()
+
+  "A foreachAsync" must {
+    "handle empty source" in assertAllStagesStopped {
+      import system.dispatcher
+      val p = Source(List.empty[Int]).runWith(Sink.foreachAsync(3)(a ⇒ Future {}))
+      Await.result(p, 200.seconds)
+    }
+
+    "be able to run elements in parallel" in assertAllStagesStopped {
+      implicit val ec = ExecutionContext.fromExecutorService(Executors.newFixedThreadPool(4))
+
+      val probe = TestProbe()
+      val latch = (1 to 4).map(_ → TestLatch(1)).toMap
+
+      val sink: Sink[Int, Future[Done]] = {
+        Sink.foreachAsync(4) { n: Int ⇒
+          Future {
+            latch(n)
+            probe.ref ! n
+          }
+        }
+      }
+
+      val p = Source(1 to 4).runWith(sink)
+
+      latch(1).countDown()
+      probe.expectMsg(1)
+      latch(2).countDown()
+      probe.expectMsg(2)
+      latch(3).countDown()
+      probe.expectMsg(3)
+      latch(4).countDown()
+      probe.expectMsg(4)
+
+      Await.result(p, 4.seconds)
+      assert(p.isCompleted)
+    }
+
+    "back-pressure upstream elements when downstream is slow" in assertAllStagesStopped {
+      import scala.concurrent.duration._
+
+      implicit val ec = ExecutionContext.fromExecutorService(Executors.newFixedThreadPool(1))
+
+      val probe = TestProbe()
+      val latch = (1 to 4).map(_ → TestLatch(1)).toMap
+
+      val sink: Sink[() ⇒ Int, Future[Done]] = {
+        Sink.foreachAsync(1) { (n: () ⇒ Int) ⇒
+          Future {
+            latch(n())
+            probe.ref ! n()
+            Thread.sleep(2000)
+          }
+        }
+      }
+
+      var oneCalled = false
+      var twoCalled = false
+      var threeCalled = false
+      var fourCalled = false
+
+      def one = { oneCalled = true; 1 }
+      def two = { twoCalled = true; 2 }
+      def three = { threeCalled = true; 3 }
+      def four = { fourCalled = true; 4 }
+
+      val p =
+        Source(List(one _, two _, three _, four _))
+          .runWith(sink)
+
+      latch(1).countDown()
+      probe.expectMsg(1)
+
+      assert(!twoCalled)
+      assert(!threeCalled)
+      assert(!fourCalled)
+
+      probe.expectNoMessage(2 seconds)
+
+      latch(2).countDown()
+      probe.expectMsg(2)
+
+      assert(!threeCalled)
+      assert(!fourCalled)
+
+      probe.expectNoMessage(2 seconds)
+
+      latch(3).countDown()
+      probe.expectMsg(3)
+
+      assert(!fourCalled)
+
+      probe.expectNoMessage(2 seconds)
+
+      latch(4).countDown()
+      probe.expectMsg(4)
+
+      Await.result(p, 4.seconds)
+      assert(p.isCompleted)
+
+      assert(oneCalled)
+      assert(twoCalled)
+      assert(threeCalled)
+      assert(fourCalled)
+    }
+  }
+
```

Comment:
Could you also take a look at the existing `SinkForeachParallelSpec` and see if some of it can be incorporated here, but implemented with `foreachAsync`.